File: server/app/services/chat/context/summary.py
"""
Summary generation and management module.

This module handles the generation and retrieval of conversation summaries.
Summaries are versioned to track the evolution of conversation context.

Summary Generation Rules:
- V < N: No summary generated (not enough messages)
- N <= V < 2N: Full summary using all messages (1 to V) with empty baseline
- V >= 2N: Incremental summary using baseline (V-N) + recent messages (V-N+1 to V)

The summary system supports:
- Recursive baseline generation when baseline is missing
- Asynchronous generation that doesn't block chat responses
- Version tracking for context assembly
"""

from sqlalchemy.orm import Session
from typing import Optional
from app.models.historical_summary import HistoricalSummary
from app.models.message import Message
from app.models.llm_config import LLMConfig
from app.services.llm import LLMService
from app.logger import logger
from langchain_core.messages import HumanMessage


class SummaryService:
    """
    Service for managing conversation summaries.
    
    This service handles:
    - Summary retrieval by version
    - Summary generation with LLM
    - Recursive baseline generation
    - Message formatting for summary prompts
    """
    
    SUMMARY_PROMPT = """You are a conversation summary assistant. Generate a new summary based on the conversation history and baseline summary.

Baseline summary (if exists):
{baseline_summary}

Recent conversation:
{recent_messages}

Generate a concise but complete summary that includes key information, decisions, and context from the conversation. The summary should help understand the background for subsequent conversations.

IMPORTANT: The summary MUST preserve the original language of the conversation.
- If the conversation is in Chinese, write the summary in Chinese.
- If the conversation is in English, write the summary in English.
- If the conversation contains multiple languages, the summary may also contain multiple languages.
- Do NOT translate between languages. Maintain information fidelity."""

    @staticmethod
    def get_summary(db: Session, session_id: int, version: int) -> Optional[HistoricalSummary]:
        """
        Get a specific summary by session ID and version.
        
        Args:
            db: Database session
            session_id: Session ID
            version: Summary version number
            
        Returns:
            HistoricalSummary if found, None otherwise
        """
        return db.query(HistoricalSummary).filter(
            HistoricalSummary.session_id == session_id,
            HistoricalSummary.version == version
        ).first()

    @staticmethod
    def get_latest_summary(db: Session, session_id: int) -> Optional[HistoricalSummary]:
        """
        Get the latest summary for a session.
        
        Used during context assembly to get the most recent summary
        available, even if it's older than the current message count.
        
        Args:
            db: Database session
            session_id: Session ID
            
        Returns:
            Latest HistoricalSummary if any exists, None otherwise
        """
        return db.query(HistoricalSummary).filter(
            HistoricalSummary.session_id == session_id
        ).order_by(HistoricalSummary.version.desc()).first()

    @staticmethod
    def get_messages_by_range(
        db: Session,
        session_id: int,
        start_seq: int,
        end_seq: int
    ) -> list[Message]:
        """
        Get messages by session-internal sequence numbers (not global IDs).
        
        Messages are ordered by their global ID, which corresponds to
        their insertion order. The sequence numbers are 1-based.
        
        Args:
            db: Database session
            session_id: Session ID
            start_seq: Start sequence number (1-based, inclusive)
            end_seq: End sequence number (inclusive)
            
        Returns:
            List of Message objects in chronological order
        """
        return db.query(Message).filter(
            Message.session_id == session_id
        ).order_by(Message.id.asc()).offset(start_seq - 1).limit(end_seq - start_seq + 1).all()

    @staticmethod
    def create_summary(
        db: Session,
        session_id: int,
        version: int,
        content: str
    ) -> HistoricalSummary:
        """
        Create and persist a new summary.
        
        Args:
            db: Database session
            session_id: Session ID
            version: Version number (equals message count at generation time)
            content: Summary content text
            
        Returns:
            Created HistoricalSummary instance
        """
        summary = HistoricalSummary(
            session_id=session_id,
            version=version,
            content=content
        )
        db.add(summary)
        db.commit()
        db.refresh(summary)
        return summary

    @staticmethod
    def format_messages_for_summary(messages: list[Message]) -> str:
        """
        Format messages for the summary prompt.
        
        Converts message objects into a human-readable format suitable
        for LLM summarization.
        
        Args:
            messages: List of Message objects
            
        Returns:
            Formatted string with role prefixes
        """
        formatted = []
        for msg in messages:
            role = "User" if msg.role == "user" else "Assistant"
            formatted.append(f"{role}: {msg.content}")
        return "\n\n".join(formatted)

    @staticmethod
    async def generate_summary(
        db: Session,
        session_id: int,
        version: int,
        llm_config: LLMConfig,
        window_size: int
    ) -> Optional[HistoricalSummary]:
        """
        Generate a summary for the specified version.
        
        This method implements the summary generation logic:
        1. Check if summary already exists (idempotent)
        2. Validate version >= window_size
        3. Determine generation strategy based on version:
           - N <= V < 2N: Full summary with all messages
           - V >= 2N: Incremental with baseline + recent messages
        4. Recursively generate missing baseline if needed
        5. Call LLM to generate summary content
        6. Persist the summary
        
        Args:
            db: Database session
            session_id: Session ID
            version: Version number (message count V)
            llm_config: LLM configuration for generation
            window_size: Summary window size (N)
            
        Returns:
            Generated HistoricalSummary if successful, None otherwise
        """
        # Check if summary already exists (idempotent)
        existing = SummaryService.get_summary(db, session_id, version)
        if existing:
            logger.info(f"Summary already exists for session {session_id}, version {version}")
            return existing

        # Validate minimum version
        if version < window_size:
            logger.info(f"Version {version} < window_size {window_size}, skipping summary generation")
            return None

        # Branch 1: N <= V < 2N - Full summary with all messages
        if version < 2 * window_size:
            messages = SummaryService.get_messages_by_range(db, session_id, 1, version)
            if not messages:
                logger.warning(f"No messages found for session {session_id}, range 1-{version}")
                return None

            messages_text = SummaryService.format_messages_for_summary(messages)
            prompt = SummaryService.SUMMARY_PROMPT.format(
                baseline_summary="(No baseline summary, this is the first summary)",
                recent_messages=messages_text
            )
            
        # Branch 2: V >= 2N - Incremental summary with baseline
        else:
            baseline_version = version - window_size
            
            # Get or recursively generate baseline summary
            baseline_summary = SummaryService.get_summary(db, session_id, baseline_version)
            if not baseline_summary:
                logger.info(f"Baseline summary not found for session {session_id}, version {baseline_version}, generating recursively...")
                baseline_summary = await SummaryService.generate_summary(
                    db, session_id, baseline_version, llm_config, window_size
                )

            baseline_text = baseline_summary.content if baseline_summary else "(No baseline summary)"

            # Get recent messages for the window
            start_seq = version - window_size + 1
            end_seq = version
            messages = SummaryService.get_messages_by_range(db, session_id, start_seq, end_seq)

            if not messages:
                logger.warning(f"No messages found for session {session_id}, seq range {start_seq}-{end_seq}")
                return None

            messages_text = SummaryService.format_messages_for_summary(messages)
            prompt = SummaryService.SUMMARY_PROMPT.format(
                baseline_summary=baseline_text,
                recent_messages=messages_text
            )

        # Generate summary using LLM
        try:
            chat_model = LLMService.create_chat_model(llm_config)
            langchain_messages = [
                HumanMessage(content=prompt)
            ]

            response = await chat_model.ainvoke(langchain_messages)
            summary_content = response.content

            # Persist the summary
            summary = SummaryService.create_summary(
                db, session_id, version, summary_content
            )

            logger.info(f"Created summary for session {session_id}, version {version}")
            return summary

        except Exception as e:
            logger.error(f"Failed to generate summary: {str(e)}", exc_info=True)
            return None

File: server/internal/service/chat/chatctx/summary.go
// Package context implements the context engineering pipeline for chat processing.
//
// This package provides the context assembly services that build the LLM message
// sequence from various context sources: summaries, narratives, retrieval results,
// user state, and task results. It matches Python's chat/context module.
package chatctx

import (
	stdctx "context"
	"fmt"

	"private-buddy-server/internal/database"
	"private-buddy-server/internal/model"
	"private-buddy-server/internal/service/llm"

	applogger "private-buddy-server/internal/logger"
)

// summaryPrompt is the LLM prompt template for conversation summarization.
// It takes two parameters: baseline_summary and recent_messages.
const summaryPrompt = `You are a conversation summary assistant. Generate a new summary based on the conversation history and baseline summary.

Baseline summary (if exists):
%s

Recent conversation:
%s

Generate a concise but complete summary that includes key information, decisions, and context from the conversation. The summary should help understand the background for subsequent conversations.

IMPORTANT: The summary MUST preserve the original language of the conversation.
- If the conversation is in Chinese, write the summary in Chinese.
- If the conversation is in English, write the summary in English.
- If the conversation contains multiple languages, the summary may also contain multiple languages.
- Do NOT translate between languages. Maintain information fidelity.`

// GenerateSummary generates a summary for the specified version.
//
// Summary Generation Rules:
//   - V < N: No summary generated (not enough messages)
//   - N <= V < 2N: Full summary using all messages (1 to V) with empty baseline
//   - V >= 2N: Incremental summary using baseline (V-N) + recent messages (V-N+1 to V)
//
// After summary content is generated, a narrative is immediately generated from
// the summary content. Both are written to the database in a single atomic
// operation, ensuring no intermediate state exists where summary exists but
// narrative is empty. This eliminates the real-time narrative generation
// bottleneck during chat processing while maintaining data consistency.
//
// Atomic write design: summary and narrative are both generated before
// writing to the database. This eliminates the intermediate state where
// summary exists but narrative is empty, avoiding race conditions during
// concurrent reads. If narrative generation fails, the entire operation
// is aborted and no record is written — the next trigger will retry.
func GenerateSummary(sessionID int64, llmConfig *model.LLMConfig, version int, windowSize int) error {
	existing := getSummary(sessionID, version)
	if existing != nil {
		applogger.L.Info("Summary already exists", "session_id", sessionID, "version", version)
		return nil
	}

	if version < windowSize {
		applogger.L.Info("Version < window_size, skipping summary generation",
			"session_id", sessionID, "version", version, "window_size", windowSize)
		return nil
	}

	var prompt string

	if version < 2*windowSize {
		messages := getMessagesByRange(sessionID, 1, version)
		if len(messages) == 0 {
			applogger.L.Warn("No messages found for session", "session_id", sessionID, "range", fmt.Sprintf("1-%d", version))
			return nil
		}

		messagesText := formatMessagesForSummary(messages)
		prompt = fmt.Sprintf(summaryPrompt, "(No baseline summary, this is the first summary)", messagesText)
	} else {
		baselineVersion := version - windowSize

		baselineSummary := getSummary(sessionID, baselineVersion)
		if baselineSummary == nil {
			applogger.L.Info("Baseline summary not found, generating recursively",
				"session_id", sessionID, "baseline_version", baselineVersion)
			if err := GenerateSummary(sessionID, llmConfig, baselineVersion, windowSize); err != nil {
				applogger.L.Error("Failed to generate baseline summary recursively",
					"session_id", sessionID, "baseline_version", baselineVersion, "error", err)
			}
			baselineSummary = getSummary(sessionID, baselineVersion)
		}

		baselineText := "(No baseline summary)"
		if baselineSummary != nil {
			baselineText = baselineSummary.Content
		}

		startSeq := version - windowSize + 1
		messages := getMessagesByRange(sessionID, startSeq, version)
		if len(messages) == 0 {
			applogger.L.Warn("No messages found for session", "session_id", sessionID, "range", fmt.Sprintf("%d-%d", startSeq, version))
			return nil
		}

		messagesText := formatMessagesForSummary(messages)
		prompt = fmt.Sprintf(summaryPrompt, baselineText, messagesText)
	}

	chatModel := llm.NewChatModelWithTemperature(llmConfig.BaseURL, llmConfig.APIKey, llmConfig.ModelID, llm.TemperatureCreative)
	summaryContent, err := chatModel.Chat(stdctx.Background(), []llm.Message{
		{Role: "user", Content: prompt},
	})
	if err != nil {
		applogger.L.Error("Summary generation LLM call failed", "session_id", sessionID, "error", err)
		return fmt.Errorf("summary generation failed: %w", err)
	}

	applogger.L.Info("Generated summary content", "session_id", sessionID, "version", version)

	narrativeResult := GenerateNarrativeFromSummary(llmConfig, summaryContent)
	if narrativeResult == "" {
		applogger.L.Error("Narrative generation failed, aborting atomic write", "session_id", sessionID, "version", version)
		return fmt.Errorf("narrative generation failed")
	}

	newSummary := model.HistoricalSummary{
		SessionID: sessionID,
		Version:   version,
		Content:   summaryContent,
		Narrative: narrativeResult,
	}
	if err := database.DB.Create(&newSummary).Error; err != nil {
		return err
	}

	applogger.L.Info("Atomically created summary+record", "session_id", sessionID, "version", version)
	return nil
}

// getSummary retrieves a specific summary by session ID and version.
func getSummary(sessionID int64, version int) *model.HistoricalSummary {
	var summary model.HistoricalSummary
	err := database.DB.Where("session_id = ? AND version = ?", sessionID, version).First(&summary).Error
	if err != nil {
		return nil
	}
	return &summary
}

// getMessagesByRange returns messages by session-internal sequence numbers (1-based, inclusive).
// Messages are ordered by their global ID, which corresponds to their insertion order.
func getMessagesByRange(sessionID int64, startSeq, endSeq int) []model.Message {
	var messages []model.Message
	database.DB.Where("session_id = ?", sessionID).
		Order("id ASC").
		Offset(startSeq - 1).
		Limit(endSeq - startSeq + 1).
		Find(&messages)
	return messages
}

// formatMessagesForSummary formats messages for the summary prompt.
// Converts message objects into a human-readable format suitable for LLM summarization.
func formatMessagesForSummary(messages []model.Message) string {
	var formatted []string
	for _, msg := range messages {
		role := "User"
		if msg.Role != "user" {
			role = "Assistant"
		}
		formatted = append(formatted, fmt.Sprintf("%s: %s", role, msg.Content))
	}
	result := ""
	for i, s := range formatted {
		if i > 0 {
			result += "\n\n"
		}
		result += s
	}
	return result
}

// GetLatestSummaryByID returns the latest summary for a session by ID.
func GetLatestSummaryByID(sessionID int64) *model.HistoricalSummary {
	var summary model.HistoricalSummary
	err := database.DB.Where("session_id = ?", sessionID).Order("version DESC").First(&summary).Error
	if err != nil {
		return nil
	}
	return &summary
}

// GenerateSummaryForSession is a shared function for triggering summary generation
// from both the API handler and ChatService, matching Python's generate_summary_task.
// It loads the session, agent, and LLM config before delegating to GenerateSummary.
func GenerateSummaryForSession(sessionID int64, version int, windowSize int) {
	var session model.Session
	if err := database.DB.First(&session, sessionID).Error; err != nil {
		applogger.L.Error("Session not found for summary generation", "session_id", sessionID, "error", err)
		return
	}

	var agent model.Agent
	if err := database.DB.First(&agent, session.AgentID).Error; err != nil {
		applogger.L.Error("Agent not found for summary generation", "agent_id", session.AgentID, "error", err)
		return
	}

	var llmConfig model.LLMConfig
	if err := database.DB.First(&llmConfig, agent.LLMConfigID).Error; err != nil {
		applogger.L.Error("LLMConfig not found for summary generation", "config_id", agent.LLMConfigID, "error", err)
		return
	}

	if err := GenerateSummary(sessionID, &llmConfig, version, windowSize); err != nil {
		applogger.L.Error("Summary generation failed", "session_id", sessionID, "error", err)
	}
}

// GetContextMessages retrieves messages for task context assembly.
// If a summary exists, only messages after the summary are returned;
// otherwise, all session messages are returned.
func GetContextMessages(sessionID int64, maxIterations int) []llm.Message {
	var summary model.HistoricalSummary
	err := database.DB.Where("session_id = ?", sessionID).Order("version DESC").First(&summary).Error

	var messages []model.Message
	if err == nil {
		database.DB.Where("session_id = ? AND id > ?", sessionID, summary.ID).
			Order("created_at ASC").Find(&messages)
	} else {
		database.DB.Where("session_id = ?", sessionID).
			Order("created_at ASC").Find(&messages)
	}

	if len(messages) > maxIterations*2 {
		messages = messages[len(messages)-maxIterations*2:]
	}

	result := make([]llm.Message, 0, len(messages))
	for _, m := range messages {
		result = append(result, llm.Message{
			Role:    m.Role,
			Content: m.Content,
		})
	}

	return result
}

// BuildSystemPrompt constructs the system prompt from agent character settings
// and cached narrative context.
func BuildSystemPrompt(agent *model.Agent, summary *model.HistoricalSummary) string {
	prompt := agent.CharacterSettings

	if summary != nil && summary.Narrative != "" {
		prompt += fmt.Sprintf("\n\n[Narrative Context]\n%s", summary.Narrative)
	}

	return prompt
}
